Sequencer class with lambda response message

DanielaMurin opened the pull request:
New Sequencer Class
Sequence number allocated according to the order that the events are received in Syncd
Passed sequence index in Syncd to function logic
The response is turned into a lambda and executed with the correct sequence order
Only added the sequence index to create/remove/set/get/bulk operations

Change-Id: Ic684138907605a2d79db9c85b84c425a5c5b0c59

shiraez committed:
Ring Buffer - Phase 1: The main thread push to the ring buffer and the second thread pop from the ring buffer and executes the function

DanielaMurin committed:
Sequencer class with lambda response message

Change-Id: Ic684138907605a2d79db9c85b84c425a5c5b0c59

DanielaMurin committed:
Add Sequencer to sendGetResponse

Change-Id: I647dc326aac995785ebc1b2d882f85fd2b3d484d

DanielaMurin committed:
Remove singleton & fix compilation

Change-Id: I001bdc0f32d81eb847157a2f7ec8bc0faa5de1a8

DanielaMurin committed:
Fix merge issues

Change-Id: Ie53906422adfc2027710ef9835dafe141448b84f

shiraez committed:
add debug info + enable thread

DanielaMurin committed:
debug single ring buffer & enable sequencer

Change-Id: I90427922c574e9a5edaa722727878febe6c4d792

shiraez committed:
add delay

DanielaMurin committed:
Add new debug file & move sequencer alloc to processEvent

Change-Id: I9b7e9de507fb3926aa2c663548e43c97d0b99daa

DanielaMurin committed:
fix reset sequencer logic

Change-Id: I380534a136ced86bb0e3e3f06f31be8cc389ec74

DanielaMurin committed:
Take out lambda execute from within lock

Change-Id: I861379abf3db49fbe3488a874f092fbeb136491d

DanielaMurin committed:
Working single ring buffer & sequencer

Change-Id: Ia215d224868aef2a5767d7b03082398ffa38e562

shiraez committed:
add loger

shiraez committed:
Merge branch 'syncd_multithread' of https://github.com/Marvell-switching/sonic-sairedis into syncd_multithread

shiraez committed:
Merge branch 'syncd_multithread' of https://github.com/Marvell-switching/sonic-sairedis into syncd_multithread

shiraez committed:
add logger to syncd + sequencer

DanielaMurin committed:
Return m_switches and m_mdioIpcServer logic

Change-Id: I24f0acbea04a99109371705b9529a0461779bfa6

shiraez committed:
fix Logger

shiraez committed:
Merge branch 'syncd_multithread' of https://github.com/Marvell-switching/sonic-sairedis into syncd_multithread

shiraez committed:
add logger.h file

DanielaMurin committed:
Update sequencer statistics/mux/logging

Change-Id: I90ef1f366c6b5f13c856d82f4aa9579c6fa9ff5b

shiraez committed:
enable flag to the logger

DanielaMurin committed:
Merge conflict fix, remove old logger code

Change-Id: I0e663cd3b7ce155298efd4f74fde86130ccf3b01

DanielaMurin committed:
Return diag shell thread, improve lambda func readability

Change-Id: Id77121b6a52a535ffda7b31c699e0b6dc9a75ae1

DanielaMurin committed:
Allocate sequence number inside functions

Change-Id: I31bc11dd855c80ae551c1cdd6b750f9c06a24947

aviramd committed:
add multiRing buffer support & send response for all ops

Fixed Problems/New Changes:

Should be Tested:

aviramd committed:
fix find operation group by string

Should be Tested:

marvellgit commented on syncd/Makefile.am:
```diff
@@ -56,7 +56,8 @@ libSyncd_a_SOURCES = \
 				ZeroMQNotificationProducer.cpp \
-				syncd_main.cpp
+				syncd_main.cpp \
+				Sequencer.cpp
```
why is ring buffer not here  ?

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+#include "Sequencer.h"
+#include "swss/logger.h"
```
add under ifdef

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+#include "Sequencer.h"
+#include "swss/logger.h"
+
+using namespace syncd;
```
why is this needed  ?

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+    std::string logMsg;
+    SequenceStatus status = FAILURE;
+ 
+    logMsg = "multithreaded: Checking for ready responses in queue... \n";
```
add under debug flag or under if def

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+        // Check if the next sequence number is in the map
+        auto seq_data = responses.find(next_seq_to_send);
+        if (seq_data == responses.end()) {
+            logMsg += "multithreaded: No next sequence found in queue \n";
```
add under debug flag or under if def

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+            status = NULL_PTR;
+        }
+        else {
+            logMsg += "multithreaded: response lambda is null \n";
```
add under debug flag or under if def

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+        // Increment the sequence number
+        ++next_seq_to_send; 
+
+        logMsg += "multithreaded: Next sequence found! Executed lambda with seq: " + std::to_string(next_seq_to_send) + " \n";
```
add under debug flag or under if def

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+    // increment the sequence number
+    current_seq++;
+
+    logMsg = "multithreaded: allocate seq num: " + std::to_string(*seq_num) + ", ";
```
add under debug flag or under if def

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+
+    // reset number to avoid overflow
+    if (current_seq >= MAX_SEQUENCE_NUMBER) {
+        logMsg += "multithreaded: Resetting allocated sequence number to avoid overflow, ";
```
add under debug flag or under if def

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+        logMsg = "multithreaded: Sequencer is full, cannot allocate sequence number" + std::to_string(current_seq);
+        LogToModuleFile("1", logMsg);
+        *seq_num = INVALID_SEQUENCE_NUMBER;
+        return BUFFER_OVERFLOW;
```
strange return status 
is this legitimate case ?
what to do ? 
1. wait untill gte free sequence
2. retunr full and have application handle 
what does ring buffre do ?

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+        }
+        
+        // Increment the number of executed tasks in sequence
+        max_num_of_executed_tasks_in_sequence++; 
```
this count the total exec in task which is also important 
for mac in sequence need to zero out every time we have and compare to max , can also have the max in last session ...

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+    } else {
+        // If the sequence is not the next to send, store it in the map
+        responses[seq] = response_lambda;
+        logMsg = "multithreaded: storing lambda with seq: " + std::to_string(seq) + ", next to send: " + std::to_string(next_seq_to_send) + " \n";
```
add under debug flag or under if def

marvellgit commented on syncd/Sequencer.h:
```diff
@@ -0,0 +1,71 @@
+    };
+    
+}
+
```
@aviramd  need to add show stats to dump ....
or print to log every x entries ...
also if detect "error" condition also  get stats ...
need to check if stats sum up correctly e.g. totatl allocate = store  + excutetd

marvellgit commented on syncd/SyncdMultipleRingBuff.h:
```diff
@@ -0,0 +1,148 @@
+#pragma once
```
????

marvellgit commented on syncd/SyncdMultipleRingBuff.h:
```diff
@@ -0,0 +1,148 @@
+#include <atomic>
+#include <mutex>
+#include <condition_variable>
+#include <map>
```
????

marvellgit commented on syncd/Sequencer.h:
```diff
@@ -0,0 +1,71 @@
+        SequenceStatus performWatchdogCheck();
+
+        // Helper function to execute all ready responses in order
+        SequenceStatus executeReadyResponses();
```
should be protected  ?

marvellgit commented on syncd/Sequencer.h:
```diff
@@ -0,0 +1,71 @@
+        // Helper function to execute all ready responses in order
+        SequenceStatus executeReadyResponses();
+
+        SequenceStatus showStatistics();
```
if it is private who can call it ??

marvellgit commented on syncd/Sequencer.h:
```diff
@@ -0,0 +1,71 @@
+
+        SequenceStatus showStatistics();
+
+        SequenceStatus clearStatistics();
```
if it is private who can call it ??

marvellgit commented on syncd/SyncdMultipleRingBuff.h:
```diff
@@ -0,0 +1,148 @@
+        template<typename DataType, int RingSize>
+        bool RingBuffer<DataType, RingSize>::IsFull()
+        {
+                return (tail + 1) % RingSize == head;
```
can be optimise by count variable

marvellgit commented on syncd/SyncdMultipleRingBuff.h:
```diff
@@ -0,0 +1,148 @@
+        template<typename DataType, int RingSize>
+        bool RingBuffer<DataType, RingSize>::IsEmpty()
+        {
+                return tail == head;
```
can be optimise by count variable

marvellgit commented on syncd/SyncdMultipleRingBuff.h:
```diff
@@ -0,0 +1,148 @@
+
+        template<typename DataType, int RingSize>
+        DataType& RingBuffer<DataType, RingSize>::HeadEntry() {
+                return buffer[head];
```
no check if not empty ?

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+        status = executeReadyResponses();
+    } else {
+        // If the sequence is not the next to send, store it in the map
+        responses[seq] = response_lambda;
```
should we use std::move ?

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+        }
+
+        // Execute the stored lambda
+        auto func = seq_data->second;
```
should we use std::move ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -26,10 +26,32 @@
 #include "swss/producertable.h"
 #include "swss/notificationconsumer.h"
 
+
```
move new down to ease compare

marvellgit commented on syncd/Syncd.h:
```diff
@@ -26,10 +26,32 @@
 
 namespace syncd
 {
+    #define ORCH_RING_SIZE 1024
```
will need to be optimise d
32 / 64  ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -72,7 +99,9 @@ namespace syncd
-                    _In_ sai_attribute_t *attr_list);
+                    _In_ sai_attribute_t *attr_list,
+                    _In_ const swss::KeyOpFieldsValuesTuple &kco,
+                    _In_ int sequenceNumber);
```
add default  ?

marvellgit replied:
or mybe add a new oveloading 
that call original will call ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -122,50 +151,71 @@ namespace syncd
+                    _In_ const swss::KeyOpFieldsValuesTuple &kco,
+                    _In_ int sequenceNumber);
+
+            sai_status_t processQuadEventTag(
```
is this still needed  ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -254,22 +307,34 @@ namespace syncd
-                    _In_ sai_attribute_t *attr_list);
+                    _In_ sai_attribute_t *attr_list,
+                    _In_ sai_common_api_t api,
+                    _In_ const swss::KeyOpFieldsValuesTuple &kco,
```
why is kco/ api added  ? 
do the exist  ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -254,22 +307,34 @@ namespace syncd
-                    _In_ const std::string& strObjectId);
+                    _In_ const std::string& strObjectId,
+                    _In_ sai_common_api_t api,
+                    _In_ const swss::KeyOpFieldsValuesTuple &kco,
```
why is kco/ api added ?
do the exist ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -254,22 +307,34 @@ namespace syncd
                     _In_ sai_object_type_t objectType,
                     _In_ const std::string& strObjectId,
-                    _In_ sai_attribute_t *attr);
+                    _In_ sai_attribute_t *attr,
```
why is kco/ api added ?
do the exist ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -254,22 +307,34 @@ namespace syncd
                     _In_ const std::string& strObjectId,
                     _In_ uint32_t attr_count,
-                    _In_ sai_attribute_t *attr_list);
+                    _In_ sai_attribute_t *attr_list,
```
why is kco/ api added ?
do the exist ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -356,10 +470,11 @@ namespace syncd
                     _In_ sai_status_t status,
                     _In_ uint32_t attr_count,
-                    _In_ sai_attribute_t *attr_list);
+                    _In_ sai_attribute_t *attr_listsendApiResponse);
```
why is name changed ? from attr_list ?

marvellgit commented on syncd/Syncd.h:
```diff
@@ -513,5 +633,212 @@ namespace syncd
+            std::map<std::string, OperationGroup> operationGroups;
+            std::atomic<bool> ring_thread_exited{false};
+
+            void getApiRingBuffer(
```
consider adding return status

marvellgit commented on syncd/Syncd.h:
```diff
@@ -513,5 +633,212 @@ namespace syncd
+                        _In_ const swss::KeyOpFieldsValuesTuple &kco,
+                        _Out_ SyncdRing*& ringBuffer);
+                        
+            void getObjectTypeByOperation(
```
consider adding return status

marvellgit commented on syncd/Syncd.h:
```diff
@@ -513,5 +633,212 @@ namespace syncd
+
+                // Declare saiObjectTypes as a member variable
+            std::vector<SaiObjectType> saiObjectTypes = {                
+                        {SAI_OBJECT_TYPE_NULL, "SAI_OBJECT_TYPE_NULL"},
```
consider using 
#define STRINGIFY(x)  x, #x

marvellgit commented on syncd/Syncd.h:
```diff
@@ -513,5 +633,212 @@ namespace syncd
+                };                 
+
+
+                std::array<int, 43> crudOperations1Enums = {
```
why use std::array
and not c array
vector / map ..... ?

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -44,6 +45,8 @@ using namespace saimeta;
 using namespace sairediscommon;
 using namespace std::placeholders;
 
+static int gdb_mode = 0;
```
still here ?

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -208,14 +212,95 @@ Syncd::Syncd(
     SWSS_LOG_NOTICE("syncd started");
 }
 
+void Syncd::popRingBuffer(SyncdRing* ringBuffer)
```
actually this is pop and excute

marvellgit replied:
move to seperat c files all ringbuffer

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -208,14 +212,95 @@ Syncd::Syncd(
+    if (!ringBuffer || ringBuffer->Started)
+	{
+		LogToModuleFile(getNameByRingBuffer(ringBuffer), "popRingBuffer return");
+        return;
```
add counter stats  ?

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -208,14 +212,95 @@ Syncd::Syncd(
 
+void Syncd::popRingBuffer(SyncdRing* ringBuffer)
+{
+    if (!ringBuffer || ringBuffer->Started)
```
ringBuffer->Started ??? should be !ringBuffer->Started ?
strange

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+ 
+    logMsg = "multithreaded: Checking for ready responses in queue... \n";
+
+    while (true) {
```
condeider adding  while not exit thread like in ring buffer

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -298,39 +383,74 @@ bool Syncd::isInitViewMode() const
     return m_asicInitViewMode && m_commandLineOptions->m_enableTempView;
 }
 
+void point(_In_ const swss::KeyOpFieldsValuesTuple &kco){
```
?still here

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -298,39 +383,74 @@ bool Syncd::isInitViewMode() const
+        getApiRingBuffer(kco, ringBuffer);
+
+        int sequenceNumber;
+        if(int seq_status = m_sequencer->allocateSequenceNumber(&sequenceNumber) != Sequencer::SUCCESS)
```
no need for if just log return status

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -298,39 +383,74 @@ bool Syncd::isInitViewMode() const
+        int sequenceNumber;
+        if(int seq_status = m_sequencer->allocateSequenceNumber(&sequenceNumber) != Sequencer::SUCCESS)
+        {
+            std::string logMsg = "Failed to allocate sequence number: " + std::to_string(seq_status);
```
add  sww log with warning

marvellgit commented on syncd/Sequencer.cpp:
```diff
@@ -0,0 +1,175 @@
+
+// Helper function to execute all ready responses in order
+// check how many consecutive responses were executed in sucession and log it
+Sequencer::SequenceStatus Sequencer::executeReadyResponses() {
```
should  not run under lock since block new ecxcute ... ? 
but need to think if we lock each iteration - two treads can run  in parallel ...

aviramd committed:
fix find ring buffer by op or objectType

Fixed Problems/New Changes:

Should be Tested:

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -446,13 +567,14 @@ sai_status_t Syncd::processAttrCapabilityQuery(
     }
 
-    m_selectableChannel->set(sai_serialize_status(status), entry, REDIS_ASIC_STATE_COMMAND_ATTR_CAPABILITY_RESPONSE);
+    sendStausResponseSequence(status, REDIS_ASIC_STATE_COMMAND_ATTR_CAPABILITY_RESPONSE, entry, sequenceNumber);
```
diff order than m_selectableChannel->need to verify

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -633,9 +773,12 @@ sai_status_t Syncd::processFdbFlush(
-
-        attr_list = vidlist.get_attr_list();
-        attr_count = vidlist.get_attr_count();
+        
```
why is this needed - in original code this was defined before  ?

marvellgit replied:
also vidlist was initialized in previous function and not moved as parameter 
now reinitilized  why ?

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -5025,6 +5407,108 @@ static void timerWatchdogCallback(
     SWSS_LOG_ERROR("main loop execution exceeded %ld ms", span/1000);
 }
 
+void Syncd::getObjectTypeByOperation(
```
very complicated :-(

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -5076,19 +5563,22 @@ void Syncd::run()
 
         if (m_commandLineOptions->m_disableExitSleep)
-            runMainLoop = false;
+            runMainLoop = true;
```
false became true ?

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -5182,11 +5672,59 @@ void Syncd::run()
             else if (sel == m_flexCounter.get())
             {
-                processFlexCounterEvent(*(swss::ConsumerTable*)sel);
+                int sequenceNumber;
```
really need to  have a s ervice or macro - too much code

marvellgit commented on syncd/Syncd.cpp:
```diff
@@ -5296,3 +5836,73 @@ syncd_restart_type_t Syncd::handleRestartQuery(
     return RequestShutdownCommandLineOptions::stringToRestartType(op);
 }
+
+void Syncd::sendStatusAndEntryResponse(
```
see priviouse note, in should be status mybe alreadt serialized entry and than command type

marvellgit reviewed (commented):
need to reduce changes in compare
replace one line with one line  ... so compare is better
also all debug crate string shpul dbe done in macro  .....

aviramd committed:
fix sendApiResponse objectCount param on processBulkQuadEventInInitViewMode

Fixed Problems/New Changes:

Should be Tested:

shiraez committed:
removing unnecessary code

aviramd committed:
Fix send responde sync for nom get actions

Fixed Problems/New Changes:

Should be Tested:

aviramd committed:
create muliple threads, cleanup, change response seq order

Fixed Problems/New Changes:

Should be Tested:

DanielaMurin committed:
Sequencer update & Cleanup

Change namespace
Add is_exit param
Fix logs
Move statistics to public
Add new statistics parmas

Change-Id: I36f9d45babc7d4ee024aa8afde44edf1615da582

DanielaMurin committed:
Sequencer update - ring buffer instead of map

Change-Id: Ib5453d04e8d1f32a6114a32df9149df11c3bb75f

aviramd committed:
fix bug when redis reads empty key

shiraez committed:
protected RedisClient + RedisSelectableChannel

DanielaMurin committed:
REVERT - Sequencer update - ring buffer instead of map

Change-Id: I7aad3550f334ae674eb93dfd66608fefb7aa0add

aviramd committed:
remove LOCK from loop wrap it instead , add thread id prints, 3 threads
